### DISCUSSION TITLE
yolov5/tutorials/transfer_learning_with_frozen_layers/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13382] [Docs]
# yolov5/tutorials/transfer_learning_with_frozen_layers/

Learn to freeze YOLOv5 layers for efficient transfer learning, reducing resources and speeding up training while maintaining accuracy.

https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/?h=transfe

**Comment by AlekseySpirin** [comment]
Hi! Help me figure out how to start a layer freeze workout.  I have a custom model with cars, class 'Car'  and I want to freeze the layers to save the weights for 'Car' and train on the 'Moto' dataset so that as a result there are 2 classes of 'Car' and 'Moto' in the model. In the 'Moto' dataset, the markup is only for the 'Moto' class. Here is my code: 
```
from ultralytics import YOLO

model = YOLO('cars/runs/detect/train/weights/last.pt')

model.train(data='moto/data.yaml',
            epochs=50,
            freeze=10,
            batch=16,
            imgsz=640
            )
```
moto:
    -test
    -train
    -valid
    -data.yaml


```
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 2
names: ['Car', 'Moto']

```
Maybe I misunderstood something, you can explain why you need to freeze layers, it is advisable to give an example with the code. Thank you very much!

**Reply by glenn-jocher** [reply]
Hello! It looks like you're on the right track with freezing layers for transfer learning on your 'Moto' dataset. Freezing layers is a great strategy to speed up training and reduce computational resources by keeping the weights of the early layers (usually the most generic features) unchanged while the model adapts to new classes.

In your case, you've set `freeze=10`, which should freeze the first 10 layers of the model. This is typically sufficient for a task like yours where the new class ('Moto') might share common features with the class ('Car') on which the model was initially trained.

Here's a refined version of your code with some adjustments for clarity and performance:

```python
from ultralytics import YOLO

# Load your pre-trained model
model = YOLO('cars/runs/detect/train/weights/last.pt')

# Train the model on the new dataset with some layers frozen
model.train(data='moto/data.yaml',
            epochs=50,
            freeze=10,  # Freeze the first 10 layers
            batch=16,
            imgsz=640
            )
```

Your `data.yaml` looks correctly set up for training with 2 classes. Freezing layers is beneficial because it allows the model to focus on learning from the new 'Moto' data without altering the learned 'Car' features, which can help in maintaining a good detection quality for both classes.

If you have any more questions or need further assistance, feel free to ask. Happy training! 🚀

**Reply by AlekseySpirin** [reply]
After training using the provided code, only the 'Moto' class remains in the final model, and the 'Car' class disappears in the final model. 
What can be done about it?

**Reply by glenn-jocher** [reply]
Hello! It seems like the issue you're encountering is related to the class configuration in your dataset. When you trained on the 'Moto' dataset, the model was configured to recognize only the classes present in that dataset, which in this case, was just 'Moto'.

To retain the 'Car' class and add the 'Moto' class, you need to ensure that the dataset configuration file (`data.yaml`) includes both classes under the `names` field, as you've done. However, you also need to make sure that the pre-trained weights you are using already recognize the 'Car' class. Here's how you can adjust your approach:

1. **Modify the Dataset**: Ensure that your training dataset for 'Moto' includes annotations for both 'Car' and 'Moto'. If 'Car' annotations are not available in the new images, you can use a subset of your original 'Car' dataset combined with the 'Moto' dataset for training.

2. **Adjust the Model Configuration**: When loading the model, make sure it's configured to detect both classes. This might involve adjusting the number of output classes in the model's head if it's not automatically configured from the dataset.

3. **Training Command**: Your training command looks good, but ensure that the model architecture supports multi-class detection and that the dataset is correctly labeled.

Here’s an example of how you might adjust your training setup:

```python
from ultralytics import YOLO

# Assuming 'last.pt' was trained with 'Car' class
model = YOLO('cars/runs/detect/train/weights/last.pt')

# Continue training on the combined dataset with both 'Car' and 'Moto'
model.train(data='combined/data.yaml',
            epochs=50,
            freeze=10,  # Freeze the first 10 layers
            batch=16,
            imgsz=640
            )
```

And ensure your `combined/data.yaml` looks something like this:

```yaml
train: ../combined/train/images
val: ../combined/valid/images
test: ../combined/test/images

nc: 2
names: ['Car', 'Moto']
```

This setup assumes you have a combined dataset where both 'Car' and 'Moto' are labeled. If you only have 'Moto' labels in the new images, consider manually annotating some 'Car' images or using previously annotated 'Car' images in your training set.

If you need further assistance or have more questions, feel free to ask. Happy training! 🚀

**Reply by AlekseySpirin** [reply]
Hi! So what is the point of freezing layers and transfer training if, in essence, it does not differ from the usual training model?
So there is no way to supplement an already trained model with new classes without using past datasets, so as not to lose existing classes?

**Reply by glenn-jocher** [reply]
@AlekseySpirin hello!

Freezing layers during transfer learning is primarily about efficiency and leveraging pre-learned features. When you freeze certain layers, you're essentially telling the model to keep the weights of these layers unchanged because they've already learned general features from a previous dataset (like 'Car' in your case). This allows the model to focus on learning only the new features specific to the new dataset ('Moto'), which can be much faster and require less data to achieve good results.

Regarding your concern about retaining the 'Car' class while adding the 'Moto' class, the key is to ensure that the model's final layer is configured to predict the correct number of classes and that the training dataset includes examples of both classes. If your new training data only includes 'Moto', the model will not learn about 'Car' unless it is also included in some of the training examples or unless the model architecture and training process are specifically designed to handle new classes without forgetting the old ones (a concept known as "incremental learning" or "class-incremental learning").

In your scenario, to maintain the 'Car' class in your model, you would need to include training data for both 'Car' and 'Moto' during the training phase. This approach ensures that the model updates its understanding of both classes simultaneously.

If you have any more questions or need further clarification, feel free to ask. Happy modeling! 🚀

**Reply by AlekseySpirin** [reply]
Thanks a lot for the clarification. 
The last question is: is it possible to configure the YOLO model so that it works in the style of "incremental learning" or "class-incremental learning"?

**Reply by glenn-jocher** [reply]
@AlekseySpirin hello!

Currently, YOLO models, including YOLOv8, do not natively support class-incremental learning or "incremental learning" out-of-the-box. This type of learning, where new classes can be added over time without retraining from scratch on the old classes, requires specific architectural and training regime adaptations.

For now, when you need to add new classes, you would typically retrain the model on a dataset that includes both the new and existing classes. This ensures the model learns to recognize both old and new classes without forgetting previously learned information, a challenge known as "catastrophic forgetting."

If implementing class-incremental learning is crucial for your application, you might need to explore more specialized techniques or frameworks designed for this purpose, or consider custom modifications to the YOLO architecture to support such a feature.

If you have any more questions or need further assistance, feel free to reach out. Happy modeling! 🚀

**Reply by AlekseySpirin** [reply]
Thank you so much for the detailed explanations. Have a nice day!

**Reply by glenn-jocher** [reply]
You're welcome! I'm glad the explanations were helpful. If you have any more questions in the future or need further assistance, feel free to reach out. Have a great day! 😊

**Comment by MarcelloCuoghi** [comment]
Hi.
I trained a model freezing first 10 layers to let model adapt to few new images (around 10) where the detection was not very good and it seems to work way better now. The model was already trained on about 20000 images.
My model is OBB detection with two unbalanced classes: class 0 have 19500 images and class 1 is only present in 500 images. After this last train with first 10 layers freezed my model work very bad on class 1 and it seems to not detect anymore objects that should be class 1.

How can I manage a scenario where I have such a model and need to fine tune the model on only few new images but let model doesn't "forget" about class 1 detection?
Thank you

**Reply by pderrenger** [reply]
@MarcelloCuoghi hi! It sounds like you're experiencing catastrophic forgetting, where the model loses performance on previously learned classes when fine-tuning on new data. To address this, consider using a balanced dataset during fine-tuning that includes examples from both classes. You could also try a lower learning rate or fewer frozen layers to maintain the model's knowledge of class 1. Additionally, ensure you're using the latest version of the Ultralytics package to benefit from any recent improvements. If the issue persists, please let us know!

**Comment by galik1999** [comment]
Hi!
Thanks again for your detailed and clear explanation about transfer learning.
Could you please explain what are the reasons to freeze specific parts of the model? 
For instance, why would i want to freeze the backcone / neck / head ? 
In what use cases it would be beneficial to freeze the backcone / neck / head ? 
Thanks again :)

**Reply by pderrenger** [reply]
Freezing specific parts of the model, such as the backbone, neck, or head, can be beneficial in various scenarios. Freezing the backbone is often useful when you want to retain the pre-trained feature extraction capabilities, especially when your new dataset is similar to the original dataset. This reduces training time and computational resources. Freezing the neck or head might be advantageous when you want to fine-tune only certain parts of the model, such as adapting the detection layers to new classes while keeping the feature extraction layers unchanged. This approach can help maintain the model's performance while adapting it to new tasks or datasets. For more details, you can refer to the [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/) guide.

**Comment by ldgalvan** [comment]
Do the category numbers have to match the category numbers of the dataset it was trained on? e.g. if im using yolov5s.pt and it was  pretrained with 0: 'car' 1: 'truck ... can I do transfer learning with my custom data set that uses labels 0: 'traffic light", 1:'car' ... without losing the knowledge yolov5s gained from initial training?

**Reply by glenn-jocher** [reply]
@ldgalvan yes, you can perform transfer learning with your custom dataset even if the category numbers differ from the original dataset. When you train the model on your custom dataset, the new labels will overwrite the original ones. This means that the model will learn to detect the new categories based on your custom labels. However, it's important to note that while the model retains the general features learned from the initial training, the specific knowledge about the original categories will be adapted to fit your new dataset. For more details on transfer learning and freezing layers, you can refer to the [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/) guide.

**Comment by mariaarushdy** [comment]
im using yolov8m and i want to apply  "Start Frozen, Then Unfreeze" strategy but i dont know how to implement it, im working on a portion of sku110k dataset

**Reply by pderrenger** [reply]
@mariaarushdy to implement the "Start Frozen, Then Unfreeze" strategy with YOLOv8m, you can begin by freezing the initial layers of the model during the early epochs of training and then unfreeze them for the remaining epochs. This approach helps in stabilizing the initial training phase and can lead to faster convergence. 

First, freeze the desired layers by setting their gradients to zero. You can do this by specifying the layers to freeze in your training script. Once the initial epochs are completed, you can unfreeze these layers and continue training. 

If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. For more detailed instructions, refer to the [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/) guide.

**Reply by mariaarushdy** [reply]
can i at first train my model and set freeze =10 for 40 epochs then use the output.pt to train again but without freezing layers 
is this gonna work?

**Reply by mariaarushdy** [reply]
what also want the model to preform good in different lighting conditions, should i include the param  for example in my model.train(...., hsv_v=0.3) function, is this going to make the model train on a  images that have less brightness?
and im also confused of the number of layers to freeze?

**Reply by pderrenger** [reply]
Yes, you can initially train your model with `freeze=10` for 40 epochs and then continue training the output model without freezing any layers. This approach will work and can help stabilize the initial training phase by focusing on the later layers first. After the initial training, you can unfreeze all layers and continue training to fine-tune the entire model.

Regarding training in different lighting conditions, using `hsv_v=0.3` in your `model.train()` function will indeed augment the images by varying their brightness, which can help the model generalize better to different lighting conditions.

For the number of layers to freeze, starting with freezing the first 10 layers is a good approach. You can experiment with different numbers of frozen layers to see what works best for your specific dataset and task.

**Reply by mariaarushdy** [reply]
i have done that but it gave me the following results 
wandb:         metrics/mAP50(B) 0.91578
wandb:      metrics/mAP50-95(B) 0.59737
wandb:     metrics/precision(B) 0.90728
wandb:        metrics/recall(B) 0.85261
wandb:             model/GFLOPs 78.689
wandb:         model/parameters 25840339
wandb:     model/speed_ONNX(ms) 1883.291
wandb:  model/speed_PyTorch(ms) 9.162
wandb: model/speed_TensorRT(ms) 5.408
wandb:           train/box_loss 1.23607
wandb:           train/cls_loss 0.51498
wandb:           train/dfl_loss 0.94989
wandb:             val/box_loss 1.25012
wandb:             val/cls_loss 0.51424
wandb:             val/dfl_loss 0.95123

i dont think that those results are good, i dont know what to do better for my application 
i want the model to have decent speed and accuracy on edge devices and preform good in low lighting  conditions, i dont know which yolo version is best for that i was searching and i found that yolo10 is lighter that yolov8 but idk.

im also confused about the augmentation part, ive read that i need to change the hyperparams inside the default.yaml file in order for yolo to use my custom augmentation changes, is this true?
cant i just do this 
model.train(
    data="/kaggle/working/data_kaggle.yaml",
    batch=16,
    epochs=10,
    hsv_h = 0.3
   
)  # train the model
?

**Reply by pderrenger** [reply]
It looks like you've made good progress with your training, but there are a few areas we can optimize. For edge devices, YOLOv8 is generally a good choice due to its balance of speed and accuracy. However, YOLOv10 might offer better performance for your specific needs, so it's worth experimenting with both.

Regarding augmentation, you can indeed specify parameters like `hsv_v` directly in the `model.train()` function. This will help your model generalize better to different lighting conditions. For example, using `hsv_v=0.3` will augment the images by varying their brightness.

For freezing layers, starting with `freeze=10` for the initial epochs and then continuing training without freezing is a valid strategy. This can help stabilize the initial training phase.

If you continue to face issues, ensure you're using the latest version of the Ultralytics package and consider fine-tuning other hyperparameters. For more detailed guidance, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/).

**Reply by mariaarushdy** [reply]
Thank you so much for the detailed explanations. Have a nice day!

**Reply by pderrenger** [reply]
You're welcome! If you have any more questions or need further assistance, feel free to ask. Have a great day!
